src/object.c

```diff
@@ -692,28 +692,35 @@ char *strEncoding(int encoding) {
 
 /* =========================== Memory introspection ========================== */
 
+static size_t sizeOfStringObject(robj *o) {
+    size_t strSize = 0;
+    if (o->encoding == OBJ_ENCODING_INT) {
+        strSize = sizeof(long long);
+    } else if (o->encoding == OBJ_ENCODING_RAW) {
+        strSize = sdsAllocSize(o->ptr)+sizeof(*o);
+    } else if (o->encoding == OBJ_ENCODING_EMBSTR) {
+        strSize = sdslen(o->ptr)+2+sizeof(*o);
+    } else {
+        serverPanic("Unknown string encoding");
+    }
+
+    return strSize;
+}
+
 /* Returns the size in bytes consumed by the key's value in RAM.
  * Note that the returned value is just an approximation, especially in the
  * case of aggregated data types where only "sample_size" elements
  * are checked and averaged to estimate the total size. */
 #define OBJ_COMPUTE_SIZE_DEF_SAMPLES 5 /* Default sample size. */
 size_t objectComputeSize(robj *o, size_t sample_size) {
-    sds ele, ele2;
+    robj *ele, *ele2;
     dict *d;
     dictIterator *di;
     struct dictEntry *de;
     size_t asize = 0, elesize = 0, samples = 0;
 
     if (o->type == OBJ_STRING) {
-        if(o->encoding == OBJ_ENCODING_INT) {
-            asize = sizeof(*o);
-        } else if(o->encoding == OBJ_ENCODING_RAW) {
-            asize = sdsAllocSize(o->ptr)+sizeof(*o);
-        } else if(o->encoding == OBJ_ENCODING_EMBSTR) {
-            asize = sdslen(o->ptr)+2+sizeof(*o);
-        } else {
-            serverPanic("Unknown string encoding");
-        }
+        sizeOfStringObject(o);
```

Comment:
The result of `sizeOfStringObject` is completely unused here. `objectComputeSize` returns `asize` at the end. You are not setting it, so for every kind of a string it will always be 0.
Needs to be:

```
asize = sizeOfStringObject(o);
```

(Haven't checked the rest of the code)